### DISCUSSION TITLE
RFC: Track total rewards earned per contract

### DISCUSSION THREAD

**Discussion by chris-ricketts** [archway-network/discussions#9] [Ideas]
It would be incredibly useful to have a query like `TotalContractRewards`, which returns the total rewards generated by a contract over it's lifetime (regardless of rewards recipient address changes).

From the perspective of interacting with the rewards module from CosmWasm contracts, it would allow a contract to be the rewards recipient of multiple contracts and know how much each contract contributed it's total earnings.

This is not possible currently as the rewards contract can only query it's total outstanding rewards from all nominator contracts.

The addition of the `TotalContractRewards` query would allow the following flow:

1. _Nominator Contract_ (NC) sets both it's owner and rewards address to be the _Rewards Contract_ (RC). Notifies RC and specifies a _collector_.
1. RC then verifies owner and rewards addresses before querying NC's `TotalContractRewards`, storing the result.
1. _Time goes by, rewards are collected from all NCs_
1. NC _collector_ comes to collect _something_ based on generated rewards, executes Collect on RC.
1. RC queries current `TotalContractRewards` comparing it to the original stored value, calculates how much _something_ to give to _collector_.

The current workaround is for the RC to create a child contract per NC to act as the rewards recipient. As well as being both less efficient gas and lines of code wise, it also degrades UX by forcing collections to be made on a per NC basis.

**Comment by zanicar** [comment]
Moved to Discussions under Ideas as this is an RFC;
